### PR TITLE
Sort alternatives alphabetically when a command is not found

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #10893 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by fabpot** [comment]
As we are listing alternatives in a lot of different places in the framework, wouldn't it make sense to change the sort for all of them?

File: src/Symfony/Component/Console/Tests/ApplicationTest.php [hunk 1]
```diff
@@ -452,6 +452,36 @@ public function testFindAlternativeNamespace()
...
+        $application->add(new \FooCommand());
+        $application->add(new \Foo1Command());
+        $application->add(new \Foo2Command());
+        $application->add(new \Foo3Command());
+
+        $expectedAlternatives = array(
+            'afoobar',
+            'afoobar1',
+            'afoobar2',
+            'foo1:bar',
```
Comment (fabpot) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: I would have expected foo1:bar to be after foo:bar as intuitively, I sort the namespace first, then the command name. Not sure if this is something we need to fix though.

Comment (iltar) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: That's probably because of the `:` when sorting

Comment (fabpot) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: I do understand that, my point is that we need to take care of that `:` when sorting like a human would do.

Comment (javiereguiluz) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: The good thing is that in real apps we probably won't have this problem. This example (`foo:`, `foo1:`, `foo3:`) is too abstract and probably our users don't use `project:, `project1:` or `app:`, `app1:` namespaces.

**Comment by fabpot** [comment]
Have you had time to have a look at the other exceptions using alternatives to see if sorting them would also make sense?

File: src/Symfony/Component/Console/Tests/ApplicationTest.php (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: fair enough

**Comment by ro0NL** [comment]
I thought of this yesterday, and it's more or less related; wouldn't it be nice to filter the command list if a command is not found.

Ie. `bin/console` vs. `bin/console partial-namespace` render the same output, but the latter only shows matching commands.

**Comment by fabpot** [comment]
@javiereguiluz Any news?

**Comment by fabpot** [comment]
@javiereguiluz Still interested in finishing this PR?

**Comment by javiereguiluz** [comment]
Yes ... but I don't know where to look for more places where this feature could be implemented. Anyone can give me a clue? Thanks!

**Comment by fabpot** [comment]
I think looking for `levenshtein` in the code base would give you all the places where we need to make the same change.

**Comment by javiereguiluz** [comment]
I've committed some changes:

* For the list of commands, I think that sorting best matches alphabetically is the most natural thing...
* ... but for the other levenshtein candidates (service ids, param names, etc.) I propose to sort them from best to worst (i.e. from lowest to highest levenshtein distance)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php [hunk 2]
```diff
@@ -68,10 +68,12 @@ public function process(ContainerBuilder $container)
                     continue;
                 }
 
-                if (false !== strpos($definedTag, $tag) || levenshtein($tag, $definedTag) <= strlen($tag) / 3) {
-                    $candidates[] = $definedTag;
+                if (false !== strpos($definedTag, $tag) || $lev = levenshtein($tag, $definedTag) <= strlen($tag) / 3) {
+                    $candidates[$definedTag] = $lev;
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: why not just `$candidates[$definedTag] = true;`? That would avoid creating the `$lev` variable that is not used anyway.

File: src/Symfony/Component/Debug/FatalErrorHandler/UndefinedMethodFatalErrorHandler.php [hunk 3]
```diff
@@ -45,12 +45,13 @@ public function handleError(array $error, FatalErrorException $exception)
         foreach ($methods as $definedMethodName) {
             $lev = levenshtein($methodName, $definedMethodName);
             if ($lev <= strlen($methodName) / 3 || false !== strpos($definedMethodName, $methodName)) {
-                $candidates[] = $definedMethodName;
+                $candidates[$definedMethodName] = $lev;
```
Comment (fabpot) [inline src/Symfony/Component/Debug/FatalErrorHandler/UndefinedMethodFatalErrorHandler.php]: same as above, `true` should be enough.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 4]
```diff
@@ -305,9 +305,11 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
                     foreach ($this->getServiceIds() as $knownId) {
                         $lev = levenshtein($id, $knownId);
                         if ($lev <= strlen($id) / 3 || false !== strpos($knownId, $id)) {
-                            $alternatives[] = $knownId;
+                            $alternatives[$knownId] = $lev;
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Container.php]: same

File: src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php [hunk 5]
```diff
@@ -77,9 +77,11 @@ public function get($name)
             foreach ($this->parameters as $key => $parameterValue) {
                 $lev = levenshtein($name, $key);
                 if ($lev <= strlen($name) / 3 || false !== strpos($key, $name)) {
-                    $alternatives[] = $key;
+                    $alternatives[$key] = $lev;
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/ParameterBag/ParameterBag.php]: here as well

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php (hunk 2 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: I'm not sure about this. `$lev` is the score ... and we need that for the two new lines added below to sort results from best to worst:

```php
asort($candidates);
$candidates = array_keys($candidates);
```

Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/UnusedTagsPass.php]: ok

**Comment by fabpot** [comment]
Tests fail on deps=low, we should find a way to fix that.

**Comment by javiereguiluz** [comment]
I can't see which is the error that makes test fail for "deps = low". Can anyone help me? Thanks!

---

Some context --> The original test was:

```
Did you mean one of these: "bar", "baz"?
```

But after this PR, that fails on most Symfony versions, so I changed it to:

```
Did you mean one of these: "baz", "bar"?
```

and now test pass ... except when using "deps = low", which fails because the expected output is again:

```
Did you mean one of these: "bar", "baz"?
```

**Comment by iltar** [comment]
@javiereguiluz my guess is that the DependencyInjection needs a higher console version for that as it seems like it's the DependencyInjection component that fails. I could be wrong though, but that's my hunch after seeing the failure.

**Comment by javiereguiluz** [comment]
After thinking about this, what if we do the following:

1) Remove the last 3 commits
2) Keep only the original feature that sorts commands alphabetically

Reason:

We already show very good alternatives because we limit the levenhstein cost to < 3, so every alternative displayed is almost as good as the other ones. The proposed changes wouldn't make the results much better.

What do you think?

**Comment by fabpot** [comment]
Make sense 👍

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Pull request merged**